Analytics Hub: Introduce time range controller

ThomazFB opened the pull request:
Summary
==========
Fix issue #8144 by introducing the Date Range selection to the Analytics Hub main view, which can generate the expected current and the previous date ranges based on the hub entry point.

Screen Capture
==========
https://user-images.githubusercontent.com/5920403/203439654-cccb0405-9477-4b60-8088-aeefefda8a9b.mp4

How to Test
==========
### Today range
1. Open the app and wait for the My Store view to load completely
2. Select the `Today` option in the My Store view and hit the `See more` button
3. Verify that the Analytics Hub view opens with the Date range selection bar showing Today vs Yesterday's information

### Week to Date range
1. Open the app and wait for the My Store view to load completely
2. Select the `This Week` option in the My Store view and hit the `See more` button
3. Verify that the Analytics Hub view opens with the Date range selection bar showing Week to Date vs the same period of last week's information

### Month to Date range
1. Open the app and wait for the My Store view to load completely
2. Select the `This Month` option in the My Store view and hit the `See more` button
3. Verify that the Analytics Hub view opens with the Date range selection bar showing Month to Date vs the same period of last month's information


### Year to Date range
1. Open the app and wait for the My Store view to load completely
2. Select the `This Year` option in the My Store view and hit the `See more` button
3. Verify that the Analytics Hub view opens with the Date range selection bar showing Year to Date vs the same period of last year's information

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Introduce extension date operations

ThomazFB committed:
Introduce AnalyticsHubTimeRange

ThomazFB committed:
Add AnalyticsHubTimeRangeTests

ThomazFB committed:
Adjust AnalyticsHubViewModel to use the time range control

ThomazFB committed:
Rename time range class to AnalyticsHubTimeRangeController

Ecarrion commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubTimeRangeController.swift:
```diff
@@ -0,0 +1,146 @@
+    let end: Date
+}
+
+public class AnalyticsHubTimeRangeController {
```
I think it is worth adding some documentation about this type. Lets aim to have our types as documented as possible! 💪 

Additionally, I see that the main purpose of this type is to generate ranges, so what about calling it `AnalyticsHubRangeGenerator` or  `AnalyticsHubRangeGenerator`. We normally reserve the word `Controller` for the `UIKit` View Controller 😁

ThomazFB replied:
Good point @Ecarrion! Let me fix that! Just one question: maybe we could call that type a `AnalyticsHubTimeRangeSet`? The type is a data set of two time ranges and their respective descriptions. If we want to use it as a generator, maybe we should avoid making the type retain a state and leave just functions to create and return the full set. WDYT?

Ecarrion replied:
I used the word `Generator` because all of the functions begin with the prefix `generate...`. 

I'm not sure I understand the `Set` idea, to me, `Set` refers to the `Set` data structure 😅

ThomazFB replied:
I see. I mentioned `set` because the interface of the type could be consumed as a set of two TimeRanges. But I see that this can lead to confusion. Let's adopt the generator word then. Thanks!

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/61188082b5b99a1f147fc4ac9e686c114855ef32 and https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/f3666968c4000757f8f027af4c44fc45909138aa

Ecarrion commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubTimeRangeController.swift:
```diff
@@ -0,0 +1,146 @@
+import Foundation
+import Yosemite
+
+struct TimeRange {
```
I think this is worth being called `AnalyticsHubTimeRange` so other members are aware that this is specific to our module & need.

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/5cb54d83dcf696381c435980ae77ad0822624bc3

Ecarrion commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubTimeRangeController.swift:
```diff
@@ -0,0 +1,146 @@
+
+public class AnalyticsHubTimeRangeController {
+
+    private let currentTimezone = TimeZone.current
```
Since we are storing the Timezone, I think it's worth using `.autoupdatingCurrent` to keep any user change.

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/f61047222656451277b695236307b96a3a235088

Ecarrion commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubTimeRangeController.swift:
```diff
@@ -0,0 +1,146 @@
+    init(selectedTimeRange: StatsTimeRangeV4,
+         currentDate: Date = Date(),
+         currentCalendar: Calendar = Calendar(identifier: .iso8601)
+    ) {
```
Style nit: I think we should have these braces in the previous line to match the rest of the method signatures style.

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/1778fc46df248a569ebdc7aaf809c3b639155d5d

Ecarrion commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubTimeRangeController.swift:
```diff
@@ -0,0 +1,146 @@
+
+    init(selectedTimeRange: StatsTimeRangeV4,
+         currentDate: Date = Date(),
+         currentCalendar: Calendar = Calendar(identifier: .iso8601)
```
Can you explain to me a little bit about why we use this specific calendar?

ThomazFB replied:
As discussed in this PR, I've added a TODO to adjust this usage later. https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/f3666968c4000757f8f027af4c44fc45909138aa

Ecarrion commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubTimeRangeController.swift:
```diff
@@ -0,0 +1,146 @@
+    private func generatePreviousTimeRangeFrom(selectionType: SelectionType) -> TimeRange {
+        switch selectionType {
+        case .today:
+            let oneDayAgo = currentCalendar.date(byAdding: .day, value: -1, to: currentDate)!
```
I think we need to avoid using `implicitly unwrapped optionals` when dealing with dates.
It's probably better to return an optional, an error, or an estimated date that won't fail/crash.

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/bd37e9c8fb283f6b412b7354ac4e99ead6d5cefb

Ecarrion commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubTimeRangeController.swift:
```diff
@@ -0,0 +1,146 @@
+
+        dateFormatter.dateFormat = "MMM d"
+        let startDateDescription = dateFormatter.string(from: timeRange.start)
+        let endDateDescription = generateEndDateDescription(timeRange: timeRange, dateFormatter: dateFormatter)
```
We aim to have all our date formats in the `DateFormatters + Woo` extension. This is to make sure we are somehow consistent on our date formats. Can we reuse some of those? If not, it's probably good to add these there!

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/861257e4131b5bac244687b0340f64abe8135437

Ecarrion reviewed (commented):
Good job! 💪 

My biggest concern would be the `!` usage which could bite us hard in production. The other suggestions are mostly about general code maintainability.

I tested the changes and they are working ok! 

A minor thing is that I found it a bit weird that we default to `time rage -> to date` and not just to `time-range`,  because it feels like the `See More` button should expand the current analytic period the user is watching.

---

Also, let's make sure that we test this with different timezones and calendars! Maybe our APAC friends can help!

ThomazFB commented:
Hey @Ecarrion, thanks for your review! Before I start addressing all your suggestions, let me start some discussions that can affect the review.

>A minor thing is that I found it a bit weird that we default to time rage -> to date and not just to time-range, because it feels like the See More button should expand the current analytic period the user is watching.

Sorry, I'm not sure if I follow what you mean here. Could you elaborate a bit more on what you're talking about?

>Also, let's make sure that we test this with different timezones and calendars! Maybe our APAC friends can help!

This is a tricky point. Right now, the calendar set in this code is the international one (ISO 8601), but it's a temporary measure. One thing to keep in mind is that our stats range should follow the same configuration from the site. The idea is to avoid causing a stats information difference between the Analytics Hub and the site stats because of device configuration.  Also, the Woo Core stats don't seem to follow different calendars besides the international one. The only configuration I could verify affecting the Woo Core stats is the week start day configuration. 

This implementation here is still missing using the site Time zone and site week start day configuration, but I'm planning to adjust this in a different task since I still need to identify where in the WCiOS I could do that.

With that said, do you think the Analytics Hub stats should follow the device calendar configuration instead of the one set on the site?

ThomazFB committed:
Merge branch 'issue/8144-create-date-range-section' into issue/8144-configure-time-range-logic

Ecarrion commented:
> Sorry, I'm not sure if I follow what you mean here. Could you elaborate a bit more on what you're talking about?

Let me illustrate with an example! I feel like if I'm looking at the monthly stats, upon tapping the "See More" CTA I should go into a "Month" range and not to a "Month to Date" range.

Nothing to change right now tho, just a comment on what felt natural to me!

> This is a tricky point. Right now, the calendar set in this code is the international one (ISO 8601), but it's a temporary measure. 

Sorry if I'm not totally aware of this, but what is the definitive measure?

> The idea is to avoid causing a stats information difference between the Analytics Hub and the site stats because of device configuration.

Now that I think about this, in order to avoid using different ranges, should we be reusing the ranges defined for the Analytics Dashboard? I see that we have a `StatsTimeRangeV4+UI` type that handles ranges. Can we reuse something from there?

> This implementation here is still missing using the site Time zone and site week start day configuration, but I'm planning to adjust this in a different task since I still need to identify where in the WCiOS I could do that.

Oh, Sorry if I missed that you have plans to change this, It was not clear to me from the PR description. 

> With that said, do you think the Analytics Hub stats should follow the device calendar configuration instead of the one set on the site?

I think we should use the same ranges we use in core, which should be the same ones we use for the dashboard, right?

ThomazFB commented:
Thanks for the clarification and elaboration on the subject, @Ecarrion!

>Let me illustrate with an example! I feel like if I'm looking at the monthly stats, upon tapping the "See More" CTA I should go into a "Month" range and not to a "Month to Date" range.

This happens because we're replicating the Time Range picker of the Woo Core stats section, the `This Month` we use in the My Store Analytics view is called `Month to Date` in the Stats range selector:

<img width="458" alt="image" src="https://user-images.githubusercontent.com/5920403/203813178-39587b73-36c3-4bd5-8a45-50326932e46d.png">

>Now that I think about this, in order to avoid using different ranges, should we be reusing the ranges defined for the Analytics Dashboard? I see that we have a StatsTimeRangeV4+UI type that handles ranges. Can we reuse something from there?

>I think we should use the same ranges we use in core, which should be the same ones we use for the dashboard, right?

We could do that, but it's important to note that we will still [add a huge amount of different ranges](https://github.com/woocommerce/woocommerce-ios/issues/8146) to the Hub, the `StatsTimeRangeV4` serves only the four ranges displayed in the My Store view, but the Hub will have to take care of different kind of ranges like `Quarter to Date`, `Last Month`, `Last Year`, `Yesterday` and `Custom`. Hence the reason why I decided to create a separate range-building logic for the Hub instead of using the My Store one.

>Sorry if I'm not totally aware of this, but what is the definitive measure?

It was my mistake not to clarify this in the task and PR descriptions. Sorry. The definitive measure would be to create a Calendar containing the site Time zone and the site's first week day configuration, so when we do the week range calculation, it's already aware of the week rules of the site, replicating the Woo Core stats correctly and avoiding being affected by the device current timezone (If we don't do this and use the device time instead, we could introduce strange scenarios like when a merchant travels to a far country and causes a sudden difference in the stats numbers from the site).

WDYT about this context? Makes sense?

Ecarrion commented:
> We could do that, but it's important to note that we will still https://github.com/woocommerce/woocommerce-ios/issues/8146 to the Hub, the StatsTimeRangeV4 serves only the four ranges displayed in the My Store view, but the Hub will have to take care of different kind of ranges like Quarter to Date, Last Month, Last Year, Yesterday and Custom. Hence the reason why I decided to create a separate range-building logic for the Hub instead of using the My Store one.

I think that is fine, and I assume that those ranges already use the correct calendar/timezone so I rather see all of those contained and maintained in the same place. 

What do you think?

ThomazFB commented:
>I think that is fine, and I assume that those ranges already use the correct calendar/timezone so I rather see all of those contained and maintained in the same place.

My main concern is that the `StatsTimeRangeV4` is an enum creating a single isolated Date, and it aims to represent the four possible selections in the My Store tabs. To make that enum capable of serving the Analytics Hub 100%, we would have to extend it to support not just four but 11 ranges, and each of these 11 must come alongside the previous range counterpart E.g.:

- Last Quarter range should go alongside the `Quarter before Last Quarter` range: `Last Quarter Jul 1 - Sep 30, 2022 compared to Apr 1 - Jun 30, 2022`

Also, the `StatsTimeRangeV4` type gets the timezone and calendar as parameters instead of choosing which should be used. I considered consuming the `StatsTimeRangeV4` to generate the ranges, but in the end, that type delegates everything to the `Date+StartAndEnd` extensions, which we're also using in the `AnalyticsHubTimeRangeController` introduced here. It feels to me that they belong to different use cases since they come in different groups of Dates and text descriptions.

That's my perspective, but please feel free to diverge if this is not the best approach.

ThomazFB committed:
Rename TimeRange to AnalyticsHubTimeRange

ThomazFB committed:
Rename AnalyticsHubTimeRangeController to AnalyticsHubTimeRangeGenerator

ThomazFB committed:
Adopt .autoupdatingCurrent as Time zone source

Ecarrion commented:
@ThomazFB I don't want to delay you more 😅 , so let's go ahead with this approach! 
Maybe add some comments to enunciate what parts are temporal and what needs to be changed!

ThomazFB committed:
Introduce analytics hub date formatters

ThomazFB committed:
Adaopt analytics hub date formatters in AnalyticsHubTimeRangeGenerator

ThomazFB committed:
Fix braces position on AnalyticsHubTimeRangeGenerator init

ThomazFB committed:
Fix hound issues

ThomazFB committed:
Merge branch 'trunk' into issue/8144-configure-time-range-logic

ThomazFB committed:
Adjust AnalyticsHubViewModel to work with new changes

ThomazFB committed:
Rename AnalyticsHubViewModel timeRangeGenerator property

ThomazFB committed:
Improve Previous period description generation

ThomazFB committed:
Exclude implicitly unwrapped optionals usage of Date+StartAndEnd functions

ThomazFB committed:
Set implicitly unwrapped optionals to the StatsTimeRangeV4+UI

ThomazFB committed:
Overhaul AnalyticsHubTimeRangeGenerator error handling

ThomazFB committed:
Adopt usage of the time ranges inside the Order stats request

ThomazFB committed:
Fix hound issues

ThomazFB committed:
Merge branch 'trunk' into issue/8144-configure-time-range-logic

ThomazFB committed:
Adjust AnalyticsHubTimeRangeGeneratorTests

ThomazFB commented:
@Ecarrion, I've adjusted the code to comply with our discussed improvements. But I had to make some changes outside of the Analytics Hub:

- I modified the `Date+StartAndEnd` functions to return `Date?` instead of forcing implicitly unwrapped optional value returns in every operation used there.

- Since the `StatsTimeRangeV4+UI` depends on those functions declared inside `Date+StartAndEnd`, I've moved the implicitly unwrapped optional usage to there. We should track an issue outside of this project to improve the My Store stats, stop using those unwrapped optionals, and avoid the DST crashes we discussed on slack. It changes nothing to the Store stats since the unwrapped optional were already there and exposes this due improvement even more.

With that said, the PR is ready for a second round.

ThomazFB committed:
Add documentation to the AnalyticsHubTimeRangeGenerator

ThomazFB committed:
fix hound issues

ThomazFB committed:
Refactor AnalyticsHubTimeRangeGenerator guard lets

Ecarrion commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubTimeRangeGenerator.swift:
```diff
@@ -0,0 +1,208 @@
+    private var _currentTimeRange: AnalyticsHubTimeRange? = nil
+    private var _previousTimeRange: AnalyticsHubTimeRange? = nil
+
+    var currentTimeRange: AnalyticsHubTimeRange {
```
Since these methods are meant to be the public API, A quick documentation for them would be useful!

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/9c7fed19e7e918780f97399c269c568f4867d063 and https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/f23614a63440b92e1b0a1622e6165441abe2dcf0

Ecarrion commented on WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubTimeRangeGenerator.swift:
```diff
@@ -0,0 +1,208 @@
+    private let currentCalendar: Calendar
+    private let selectionType: SelectionType
+
+    private var _currentTimeRange: AnalyticsHubTimeRange? = nil
+    private var _previousTimeRange: AnalyticsHubTimeRange? = nil
```
Style Nit: Not super important but on `Swift` is not common to name private/hidden properties with an underscore.

A more common pattern is to store them as private and then expose throwing functions for them!

ThomazFB replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/9c7fed19e7e918780f97399c269c568f4867d063 and https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/f23614a63440b92e1b0a1622e6165441abe2dcf0

Ecarrion reviewed (approved):
Thanks for taking the time to address my comments @ThomazFB! 🙇 

I left some other minor style comments but we should be good to go!

ThomazFB committed:
Improve AnalyticsHubTimeRangeGenerator API documentation and disposition

ThomazFB committed:
Add documentation over description generation functions

ThomazFB commented:
Thank you @Ecarrion for all your effort in reviewing this big PR and long discussion! 🙇 I've addressed your suggestions!
